### PR TITLE
Implement market order trades, structure league+order+portfolio

### LINKED ISSUE #36: [LAPI] Make trades endpoint
Make endpoints for all trading actions, which include buying and selling equities.

See #21  for front end implementation.

### PR DESCRIPTION
**Pull request by avidlaud**
Resolves #36

The structure for portfolios is set - still working on developing an endpoint to provide portfolios.
There is now basic functionality for market orders:

- Checks valid ticker with `tradeValidation.ticker` middleware
- Checks gathers current market price and checks that the user has enough cash
- Adds (executed) order to user's portfolio
- Creates/updates holdings entry for target ticker
- Subtracts available cash from portfolio

Available through the `/trade/submit` endpoint on the league service.

### Sending an invalid ticker:

image

### Insufficient cash:

image

## Successful trade - new holding:

Purchasing two shares of twtr:
image

## Successful trade - increasing position:

Purchasing three more shares of twtr:
image

### CODE REVIEW

File: app/services/league/controllers/leagueController.js [hunk 1]
```diff
@@ -1,24 +1,50 @@
...
-    await User.findOneAndUpdate(
-        { username: user },
-        { $addToSet: { leagues: leagueID } },
-        {},
-        (err) => {
-            if (err) throw err;
-        },
-    ),
-]);
+    );
```
Comment (kjprajapati18) [inline app/services/league/controllers/leagueController.js:11]: One of the middlewares already ensures that the league exists, so idk if this is necessary

Comment (kjprajapati18) [inline app/services/league/controllers/leagueController.js:11]: One of the middlewares already ensures that the league exists, so idk if this is necessary

File: app/services/league/controllers/leagueController.js [hunk 2]
```diff
@@ -1,24 +1,50 @@
...
+                        cash: league.settings.balance,
+                        netWorth: league.settings.balance,
+                    }),
+                },
+            },
+            {},
+            (err) => {
+                if (err) throw err;
+            },
+        ),
```
Comment (kjprajapati18) [inline app/services/league/controllers/leagueController.js:37]: We should combine this into one request if possible

File: app/services/league/controllers/leagueController.js [hunk 3]
```diff
@@ -88,6 +114,14 @@ exports.leaveLeague = async (req, res) => {
...
             },
         ),
+        await League.findOneAndUpdate(
+            { 'portfolioList.owner': username },
+            { $pull: { portfolioList: { owner: username } } },
+            { new: true },
+            (err) => {
+                if (err) throw err;
+            },
+        ),
```
Comment (kjprajapati18) [inline app/services/league/controllers/leagueController.js:124]: Combine with above request from lines 101 to 108

File: app/services/league/controllers/tradeController.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        let order;
+        let userPortfolio;
+        let currPrice;
+        const { username } = res.locals;
+        const selectedLeague = await League.findOne(
+            { leagueName: req.body.leagueName },
+            (err) => {
+                if (err) throw err;
+            },
+        );
```
Comment (kjprajapati18) [inline app/services/league/controllers/tradeController.js:28]: You can use the findValidLeague middleware to find the league beforehand. Then simply access it through res.locals.league

File: app/services/league/controllers/leagueController.js (hunk 1 above)
Comment (avidlaud) [reply app/services/league/controllers/leagueController.js:11]: Done

File: app/services/league/controllers/leagueController.js (hunk 2 above)
Comment (avidlaud) [reply app/services/league/controllers/leagueController.js:37]: Done

File: app/services/league/controllers/leagueController.js (hunk 3 above)
Comment (avidlaud) [reply app/services/league/controllers/leagueController.js:124]: I don't know if this is possible

File: app/services/league/controllers/tradeController.js (hunk 4 above)
Comment (avidlaud) [reply app/services/league/controllers/tradeController.js:28]: Done

**Review by JScheire — APPROVED** [review]

**Review by kjprajapati18 — APPROVED** [review]

**Review by RiyaTayal20 — APPROVED** [review]

**Pull request merged**